### PR TITLE
[DO NOT MERGE YET] [Enhancement] [Infrastructure] Introduce "profile-stats.py" helper

### PR DESCRIPTION
**Pull request by iankko**
On the SSG mailing list we are often asked for timeframe when specific profile will be completed.
This is hard to estimate without having information, how many OVAL checks and remediations
are missing for particular profile.

Yet different use case where this might be useful - SSG contributors are sometimes wondering
what they could help with. But we are missing information for them where to start. This tool
provides candidates for the rules, for which OVAL checks and remediations can be safely
written for.

Will provide sample output in next comment.

Please review.

Thank you, Jan

### CODE REVIEW

**Comment by iankko** [comment]
Some sample outputs (full stats of existing profile):
```
$ ./profile_stats.py /tmp/github/scap-security-guide/RHEL/7/output/ssg-rhel7-xccdf.xml --profile stig-rhel7-server-upstream --full-stats
* Statistics of 'stig-rhel7-server-upstream' profile:
** Count of rules: 181
** Count of ovals: 173 [95% complete]
** Count of fixes: 137 [75% complete]
** Rules of 'stig-rhel7-server-upstream' profile missing OVAL:
   bios_assign_password                            wireless_disable_in_bios                     
   bios_disable_usb_boot                           install_hids                                 
   bios_enable_execution_restrictions              encrypt_partitions                           
   account_temp_expire_date                     
   libreswan_approved_tunnels                   
** Rules of 'stig-rhel7-server-upstream' profile missing remediation:
   installed_OS_is_certified                    
   accounts_password_all_shadowed                  bootloader_password                          
   bios_assign_password                            wireless_disable_in_bios                     
   wireless_disable_interfaces                     bios_disable_usb_boot                        
   selinux_confinement_of_daemons               
   selinux_all_devicefiles_labeled                 gid_passwd_group_same                        
   file_permissions_sshd_pub_key                
   file_permissions_sshd_private_key            
   file_permissions_ungroupowned                   no_files_unowned_by_user                     
   rpm_verify_hashes                               install_hids                                 
   install_antivirus                            
   install_PAE_kernel_on_x86-32                 
   bios_enable_execution_restrictions              accounts_tmout                               
   set_password_hashing_algorithm_libuserconf      encrypt_partitions                           
   partition_for_tmp                            
   partition_for_var_log_audit                     partition_for_var                            
   partition_for_home                           
   accounts_password_pam_maxclassrepeat            account_temp_expire_date                     
   ftp_present_banner                           
   sysctl_net_ipv6_conf_all_accept_source_route 
   network_disable_ddns_interfaces              
   libreswan_approved_tunnels                      rsyslog_remote_loghost                       
   sshd_disable_compression                        sshd_disable_gssapi_auth                     
   sshd_disable_kerb_auth                          sshd_enable_strictmodes                      
   sshd_use_priv_separation                        xwindows_runlevel_setting                    
   package_xorg-x11-server-common_removed          enable_x11_forwarding                        
   mount_option_krb_sec_remote_filesystems      
   use_kerberos_security_all_exports            
   snmpd_not_default_password                   


```

Attempt to obtain stats on non-existing profile:
```
$ ./profile_stats.py /tmp/github/scap-security-guide/RHEL/7/output/ssg-rhel7-xccdf.xml --profile pcidss
No such profile "pcidss" found in the benchmark!
* Available profiles:
** standard
** pci-dss
** C2S
** rht-ccp
** common
** stig-rhel7-workstation-upstream
** stig-rhel7-server-gui-upstream
** stig-rhel7-server-upstream
** ospp-rhel7-server
** nist-cl-il-al
** cjis-rhel7-server
```

And base stats (without listed OVAL and remediation IDs) for proper ```pci-dss``` RHEL-7 profile:
```
$ ./profile_stats.py /tmp/github/scap-security-guide/RHEL/7/output/ssg-rhel7-xccdf.xml --profile pci-dss
* Statistics of 'pci-dss' profile:
** Count of rules: 94
** Count of ovals: 93 [98% complete]
** Count of fixes: 85 [90% complete]


```

**Comment by shawndwells** [comment]
Can you add missing CCEs?

**Comment by shawndwells** [comment]
p.s. this is great!

File: shared/utils/profile_stats.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+console_width = 80
+
+class rule_stats:
+    def __init__(self, rid = None, roval = None, rfix = None):
+        self.dict = { 'id' : rid,
+                      'oval' : roval,
+                      'fix' : rfix }
+
+
+class xccdf_benchmark:
```
Comment (mpreisler) [inline shared/utils/profile_stats.py]: don't use old style classes in Python, use new style instead

File: shared/utils/profile_stats.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+
+def usage():
+    print("%s" % script_usage)
+    sys.exit(1)
+
+def main():
+
+    if len(sys.argv) < 2:
+        usage()
```
Comment (mpreisler) [inline shared/utils/profile_stats.py]: why not `argparse` or `optparse`?

**Comment by mpreisler** [comment]
Yes, this is awesome. Extra brownie points for some sort of machine readable output that we can generate using Jenkins and publish a small HTML file with graphs and stuff.

**Comment by iankko** [comment]
@shawndwells 
> Can you add missing CCEs? 

Added. Using --full-stats will now print out info also about possibly not-assigned CCEs for rule.

**Comment by iankko** [comment]
@mpreisler 
> don't use old style classes in Python, use new style instead

Fixed. 

@mpreisler 
> why not argparse or optparse?

Wasn't initially used because I remember the situation in ```count_oval_objects.py``` helper, this was rewritten from using config parser to use self handling of ```sys.argv``` (to ensure python-2.6.X to python-3.5.X compatibility). But since this explicitly requires ```python2``` rewritten to use ```optparse``` now (seemed more reasonable to me than ```argparse``` that got introduced just in 2.7 version. While ```optparse``` is going to be deprecated in future Python versions, it will still work correctly on python-2.6 [RHEL-6] and python-2.7 case [RHEL-7]).

**Comment by iankko** [comment]
@mpreisler 
> Yes, this is awesome. Extra brownie points for some sort of machine readable output that we can generate using Jenkins and publish a small HTML file with graphs and stuff.

Yet one check due this one -- can you clarify format of such machine readable output? Would be --format json preferred. Or should it produce such HTML file itself (when --html option is supplied).

If the former, would it be sufficient to dump the dictionary with stats in json format, if --format json was supplied? If the latter, would be enough to dump the content of the stats dictionary as HTML page? Or such HTML page should be created by external tool? (maybe here would help if you could specify what kind of graphs should be possible to achieve from it).

Thanks, Jan

**Comment by mpreisler** [comment]
@iankko We discussed this today. Just a small recap. Whether it's a json file or HTML file is an implementation detail. I would go with the json file because it's easier at this stage and we could quite easily generate a bunch of those files for various commits and then use them to create graphs and HTML pages. That would help us show how SCAP Security Guide is progressing over time.

File: shared/utils/profile_stats.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+class rule_stats(object):
+    def __init__(self, rid = None, roval = None, rfix = None, rcce = None):
+        self.dict = { 'id' : rid,
+                      'oval' : roval,
+                      'fix' : rfix,
+                      'cce' : rcce }
+
+
+class xccdf_benchmark(object):
```
Comment (mpreisler) [inline shared/utils/profile_stats.py]: Please rename to CamelCase to follow PEP8.

File: shared/utils/profile_stats.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+import re
+import sys
+
+xccdf_ns = "http://checklists.nist.gov/xccdf/1.1"
+oval_ns = "http://oval.mitre.org/XMLSchema/oval-definitions-5"
+rem_system = "urn:xccdf:fix:script:sh"
+cce_system = "https://nvd.nist.gov/cce/index.cfm"
+console_width = 80
+
+class rule_stats(object):
```
Comment (mpreisler) [inline shared/utils/profile_stats.py]: Please rename to CamelCase to follow PEP8.

**Comment by mpreisler** [comment]
Decided to merge this, I will fix PEP8 myself.

**Comment by mpreisler** [comment]
https://github.com/OpenSCAP/scap-security-guide/commit/b2ecab529a99cb3322c7dc89546d9b6a9466204e
https://github.com/OpenSCAP/scap-security-guide/commit/6d1d4f4005ab0cf6727484e575b8fbe21eca2dd2

**Comment by iankko** [comment]
@mpreisler 
This wasn't very helpful for me. Wanted to enhance it with https://github.com/iankko/scap-security-guide/commit/9c9b55c6aabca94bed5daaf996649e7ef1db1bf8 change yet (so it would show also implemented OVALs, remediations, and CCEs).

Nevermind, I will try to apply that one on top of current master.

**Comment by mpreisler** [comment]
Sorry about that, @iankko 
There was no activity here for 10 days so I assumed I'd be helping you by fixing the nitpicks myself.

**Pull request merged**